### PR TITLE
gh-94673: Add Per-Interpreter tp_subclasses for Static Builtin Types

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
* Issue: gh-94673

### CODE REVIEW

File: Objects/structseq.c [hunk 1]
```diff
@@ -580,7 +580,8 @@ _PyStructSequence_FiniType(PyTypeObject *type)
     assert(type->tp_base == &PyTuple_Type);
 
     // Cannot delete a type if it still has subclasses
-    if (type->tp_subclasses != NULL) {
+    if (_PyType_HasSubclasses(type)) {
+        PyErr_Clear();
```
Comment (kumaraditya303) [inline Objects/structseq.c]: Where this exception is raised?

File: Objects/typeobject.c [hunk 2]
```diff
@@ -4398,6 +4419,30 @@ type_dealloc(PyTypeObject *type)
...
+
+int
+_PyType_HasSubclasses(PyTypeObject *self)
+{
+    if (self->tp_flags & _Py_TPFLAGS_STATIC_BUILTIN &&
+            _PyStaticType_GetState(self) == NULL) {
+        return 0;
+    }
+    if (lookup_subclasses(self) == NULL) {
+        return PyErr_Occurred() ? -1 : 0;
```
Comment (kumaraditya303) [inline Objects/typeobject.c]: same question

**Review by kumaraditya303 — APPROVED** [review]
LGTM, I have left some questions. 

FWIW, we could also get rid of the new type flag and use `tp_subclasses` as a tagged pointer and steal one bit to check if the `tp_subclasses` is per interpreter or local, I'll probably do that but for now this is fine.

File: Objects/typeobject.c [hunk 3]
```diff
@@ -4300,10 +4307,14 @@ type_dealloc_common(PyTypeObject *type)
...
 
+static void clear_subclasses(PyTypeObject *self);
+
 static void
 clear_static_tp_subclasses(PyTypeObject *type)
 {
-    if (type->tp_subclasses == NULL) {
+    PyObject *subclasses = lookup_subclasses(type);
+    if (subclasses == NULL) {
+        PyErr_Clear();
```
Comment (kumaraditya303) [inline Objects/typeobject.c]: same question

**Comment by encukou** [comment]
I'm afraid this isn't backwards compatible. [The `tp_subclasses` docs](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_subclasses) do say it's internal only, but do currently guarantee that it's a list, so IMO it's valid for API users to treat it as a `PyObject*`, and expect exceptions (rather than crashes).

Consider:
* Seeking a SC exception from Python's backcompat policy
* Renaming the field to `_tp_subclasses`, so anyone currently using it gets a compiler error
* Adding the change to What's New so people know what's up
* Removing the field from docs
* Getting a PEP accepted for changes like this, to find unintended consequences

**Comment by ericsnowcurrently** [comment]
FWIW, `tp_subclasses` is already a dict and probably has been for quite a while.

**Comment by kumaraditya303** [comment]
>  but do currently guarantee that it's a list,

The current implementation uses a dictionary to store subclasses as a mapping of `PyLongObject *` to subclass. On the Python level it is a list but not at C level. 

> so IMO it's valid for API users to treat it as a PyObject*, and expect exceptions (rather than crashes).

Why do you expect it to raise than exception when you dereference a pointer which is clearly documented to be a internal implementation detail?

**Comment by encukou** [comment]
I'm not saying this is a bad change. I'm asking to be mindful of the possible effects on code other than CPython that might look at this field.

> FWIW, tp_subclasses is already a dict and probably has been for quite a while.

Yes, the docs are wrong. I don't think that's a free pass make this an invalid `PyObject*` pointer, though.
Bad docs just mean that people will experiment, or read the source, to see what things do.

> Why do you expect it to raise an exception when you dereference a pointer which is clearly documented to be a internal implementation detail?

Well, I do expect to be able to inspect internal details, e.g. for debugging.

**Comment by kumaraditya303** [comment]
@encukou Let's continue the discussion on discourse.

**Comment by ericsnowcurrently** [comment]
Hey, the feedback is super helpful!  Thanks!

I'll go ahead and add a whatsnew entry about the special-casing of `tp_subclasses` for static builtin types.

**Comment by ericsnowcurrently** [comment]
Making `tp_sublcasses` `void *` is a good idea.

File: Objects/structseq.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Objects/structseq.c]: Good catch.  `lookup_subclasses()` could fail before.  Now it can't.  I forgot to update all the call sites.  Now I have.

**Comment by encukou** [comment]
Thanks!
I left some more suggestions on the docs in https://github.com/ericsnowcurrently/cpython/pull/31

**Pull request merged**